lib/processors/organize-toc.mjs

```diff
@@ -0,0 +1,250 @@
+    return 1;
+  }
+
+  // Fall back on launch stage.
+  if (
+    LAUNCH_STAGE_SORT_ORDER.indexOf(aMatch.groups.stage) <
+    LAUNCH_STAGE_SORT_ORDER.indexOf(bMatch.groups.stage)
+  ) {
+    return -1;
+  }
+  if (
+    LAUNCH_STAGE_SORT_ORDER.indexOf(aMatch.groups.stage) >
+    LAUNCH_STAGE_SORT_ORDER.indexOf(bMatch.groups.stage)
+  ) {
+    return 1;
+  }
+
+  // Finally, try the version number associated with the launch stage.
+  if (aMatch.groups.stageVersion < bMatch.groups.stageVersion) {
+    return -1;
+  }
+  if (aMatch.groups.stageVersion > bMatch.groups.stageVersion) {
+    return 1;
+  }
+
+  return 0;
+}
+
+// Sorts lexicographically by the name of each TOC item.
+function tocItemSorter(a, b) {
+  if (a.name < b.name) {
+    return -1;
+  }
+  if (a.name > b.name) {
+    return 1;
+  }
+  return 0;
+}
+
+class TocProcessor {
+  constructor(metadata, obj) {
+    const apiModel = new ApiModel();
+    const apiPackage = apiModel.loadPackage(
+      join(metadata.tmpDir, `${metadata.packageShortName}.api.json`)
+    );
+
+    this._obj = obj;
+    // We normally check the API model to determine which versions are exported.
+    // However, we also allow tests to specify the available versions in the
+    // `metadata` object.
+    this._versions = metadata._versions ?? findExportedVersions(apiPackage);
+
+    // The TOC should have only one main item, which is the library.
+    this._mainItem = obj.items.find(
+      item => item.uid === `${metadata.packageInfo.name}!`
+    );
+    // Remove the main item's children. We'll add them back later, after we
+    // reorganize them. We store them in an `items` property so that we can
+    // recursively iterate over that property.
+    this._items = {
+      items: this._mainItem.items.splice(0),
+    };
+
+    this.noKind = [];
+    Object.keys(TOC_KINDS).forEach(kind => {
+      this[kind] = {
+        // Maps namespaces to the items they contain.
+        namespaces: {},
+        // Items with no namespace.
+        root: [],
+      };
+    });
+  }
+
+  _categorizeItems() {
+    recursiveForEach(this._items, 'items', item => {
+      let kind = item.uid?.match(KIND_REGEXP);
+
+      if (kind) {
+        // Use the captured value as the kind.
+        kind = kind[1];
+      }
+      // Remove ` (Class)` from titles like `FooBar (Class)`.
+      item.name = item.name.replace(TITLE_KIND_REGEXP, '');
+
+      // The package's top-level exported classes get special treatment. To
+      // confirm that the item represents a top-level exported class, we must
+      // also check that the item is associated with the most recent exported
+      // API version.
+      if (this._isNewestExportedVersion(item)) {
+        this._mainItem.items.push(item);
+      }
+      // Don't mess with items that have UIDs like `@google-cloud/foo!`, or a
+      // kind that we don't recognize.
+      else if (!kind || !Object.prototype.hasOwnProperty.call(this, kind)) {
```

Comment:
If I'm understanding correctly, this is not expected to happen, right? I think this behavior is fine, but I think it'd be helpful if we log cases when we see this.

Reply:
Items with UIDs like `@google-cloud/foo!` are normal and expected. Items with unrecognized kinds are not.

Updated to log a warning only for items with unrecognized kinds.